bftengine/src/bcstatetransfer/BCStateTran.cpp

```diff
@@ -2333,56 +2329,57 @@ void BCStateTran::reportCollectingStatus(const uint64_t firstRequiredBlock,
+  bj.addKv("blocksLeft", (nextRequiredBlock_ - firstRequiredBlock));
+  bj.addKv("cycle", cycleCounter_);
+  bj.addKv("elapsedTime", std::to_string(blocks_overall_r.elapsed_time_ms_) + " ms");
+  bj.addKv("collected",
+           std::to_string(blocks_overall_r.num_processed_items_) + " blk & " +
+               std::to_string(bytes_overall_r.num_processed_items_) + " B");
+  bj.addKv("throughput",
+           std::to_string(blocks_overall_r.throughput_) + " blk/s & " + std::to_string(bytes_overall_r.throughput_) +
+               " B/s");
+  bj.endNested();
 
   if (config_.gettingMissingBlocksSummaryWindowSize > 0) {
     auto blocks_win_r = blocks_collected_.getPrevWinResults();
     auto bytes_win_r = bytes_collected_.getPrevWinResults();
     auto prev_win_index = blocks_collected_.getPrevWinIndex();
 
-    nested_data.insert(toPair("index", prev_win_index));
-    nested_data.insert(toPair("elapsedTime", std::to_string(blocks_win_r.elapsed_time_ms_) + " ms"));
-    nested_data.insert(toPair("collected",
-                              std::to_string(blocks_win_r.num_processed_items_) + " blk & " +
-                                  std::to_string(bytes_win_r.num_processed_items_) + " B"));
-    nested_data.insert(toPair(
-        "throughput",
-        std::to_string(blocks_win_r.throughput_) + " blk/s & " + std::to_string(bytes_win_r.throughput_) + " B/s"));
-    result.insert(
-        toPair("lastWindow", concordUtils::kvContainerToJson(nested_data, [](const auto &arg) { return arg; })));
-    nested_data.clear();
-  }
-
-  nested_data.insert(toPair("lastStored", psd_->getLastStoredCheckpoint()));
-  nested_nested_data.insert(toPair("checkpointNum", fetched_cp.checkpointNum));
-  nested_nested_data.insert(toPair("lastBlock", fetched_cp.lastBlock));
-  nested_data.insert(
-      toPair("beingFetched", concordUtils::kvContainerToJson(nested_nested_data, [](const auto &arg) { return arg; })));
-  result.insert(
-      toPair("checkpointInfo", concordUtils::kvContainerToJson(nested_data, [](const auto &arg) { return arg; })));
-
-  oss << concordUtils::kContainerToJson(result);
-  return oss.str().c_str();
+    bj.startNested("lastWindow");
+    bj.addKv("index", prev_win_index);
+    bj.addKv("elapsedTime", std::to_string(blocks_win_r.elapsed_time_ms_) + " ms");
+    bj.addKv("collected",
+             std::to_string(blocks_win_r.num_processed_items_) + " blk & " +
+                 std::to_string(bytes_win_r.num_processed_items_) + " B");
+    bj.addKv("throughput",
+             std::to_string(blocks_win_r.throughput_) + " blk/s & " + std::to_string(bytes_win_r.throughput_) + " B/s");
+    bj.endNested();
+  }
+
+  bj.startNested("checkpointInfo");
+
+  bj.addKv("lastStored", psd_->getLastStoredCheckpoint());
+
+  bj.startNested("beingFetched");
+  bj.addKv("checkpointNum", fetched_cp.checkpointNum);
+  bj.addKv("lastBlock", fetched_cp.lastBlock);
+  bj.endNested();
+
+  bj.endNested();
+
+  bj.endJson();
```

Comment:
I think of the `startJson()`/`startNested()` and `endJson()`/`endNested()` as the "curly braces" for the JSON clauses.  If the developer forgets to add `endNested()` in the middle of the JSON construction, the nested JSON clause will (incorrectly) continue.

I don't think we can get rid of `endJson()` and do the equivalent in the destructor, since `getJson()` needs to be called before the `BuildJson` destructor is called.